dpcpp/preconditioner/jacobi_kernels.dp.cpp

```diff
@@ -33,212 +33,437 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/preconditioner/jacobi_kernels.hpp"
 
 
-#include <algorithm>
-#include <cmath>
-#include <iterator>
-#include <numeric>
-#include <vector>
+#include <dpcpp/preconditioner/jacobi_common.hpp>
 
 
 #include <CL/sycl.hpp>
 
 
 #include <ginkgo/core/base/exception_helpers.hpp>
-#include <ginkgo/core/base/math.hpp>
-#include <ginkgo/core/matrix/csr.hpp>
-#include <ginkgo/core/matrix/dense.hpp>
 
 
-#include "core/base/allocator.hpp"
 #include "core/base/extended_float.hpp"
 #include "core/preconditioner/jacobi_utils.hpp"
-#include "dpcpp/components/matrix_operations.dp.hpp"
+#include "core/synthesizer/implementation_selection.hpp"
+#include "dpcpp/base/config.hpp"
+#include "dpcpp/base/dim3.dp.hpp"
+#include "dpcpp/components/cooperative_groups.dp.hpp"
+#include "dpcpp/components/thread_ids.dp.hpp"
 
 
 namespace gko {
 namespace kernels {
 namespace dpcpp {
 /**
  * @brief The Jacobi preconditioner namespace.
- *
+ * @ref Jacobi
  * @ingroup jacobi
  */
 namespace jacobi {
-
-
-void initialize_precisions(std::shared_ptr<const DpcppExecutor> exec,
-                           const array<precision_reduction>& source,
-                           array<precision_reduction>& precisions)
-    GKO_NOT_IMPLEMENTED;
-
-
 namespace {
 
 
-template <typename IndexType>
-inline bool has_same_nonzero_pattern(
-    const IndexType* prev_row_ptr, const IndexType* curr_row_ptr,
-    const IndexType* next_row_ptr) GKO_NOT_IMPLEMENTED;
-
-
-template <typename ValueType, typename IndexType>
-size_type find_natural_blocks(const matrix::Csr<ValueType, IndexType>* mtx,
-                              uint32 max_block_size,
-                              IndexType* block_ptrs) GKO_NOT_IMPLEMENTED;
+// a total of 8 32-subgroup (256 threads)
+constexpr int default_num_warps = 8;
```

Comment:
Maybe base this off `GKO_DPCPP_JACOBI_BLOCK_SIZES` (or the max of it), to reduce the magic numbers a bit? We also have `config:::max_block_size` which could be used instead of the `256`?